lib/WUI/nhttp/req_parser.cpp

```diff
@@ -0,0 +1,177 @@
+using std::get_if;
+using std::holds_alternative;
+using std::monostate;
+using std::string_view;
+
+namespace {
+
+const Automaton http_request(nhttp::parser::request::paths, nhttp::parser::request::transitions, nhttp::parser::request::states);
+
+}
+
+namespace nhttp::handler {
+
+RequestParser::RequestParser(const Server &server)
+    : Execution(&http_request)
+    , server(&server) {}
+
+ExecutionControl RequestParser::event(Event event) {
+    switch (event.leaving_state) {
+    case Names::MethodGet:
+        method = Method::Get;
+        return ExecutionControl::Continue;
+    case Names::MethodHead:
+        method = Method::Head;
+        return ExecutionControl::Continue;
+    case Names::MethodPost:
+        method = Method::Post;
+        return ExecutionControl::Continue;
+    case Names::MethodDelete:
+        method = Method::Delete;
+        return ExecutionControl::Continue;
+    case Names::MethodPut:
+        method = Method::Put;
+        return ExecutionControl::Continue;
+    case Names::MethodUnknown:
+        method = Method::Unknown;
+        return ExecutionControl::Continue;
+    }
+
+    switch (event.entering_state) {
+    case Names::Url:
+        if (url_size < url.size()) {
+            url[url_size++] = event.payload;
+            return ExecutionControl::Continue;
+        } else {
+            error_code = Status::UriTooLong;
+            return ExecutionControl::Continue;
+        }
+    case Names::Boundary:
+        if (boundary_size < boundary_buff.size()) {
+            boundary_buff[boundary_size++] = event.payload;
+            return ExecutionControl::Continue;
+        } else {
+            error_code = Status::RequestHeaderFieldsTooLarge;
+            return ExecutionControl::Continue;
+        }
+    case Names::XApiKey:
+        // No API key -> can't authenticate.
+        if (!api_key) {
+            auth_status = false;
+        }
+        // The first character, just starting.
+        if (holds_alternative<monostate>(auth_status)) {
+            auth_status = static_cast<uint8_t>(0);
+        }
+        if (uint8_t *idx = get_if<uint8_t>(&auth_status); idx != nullptr) {
+            const size_t key_len = strlen(api_key);
+            if (*idx >= key_len || api_key[*idx] != event.payload) {
+                auth_status = false;
+            } else if (++*idx == key_len) {
+                auth_status = true;
+            }
+        } else {
+            auth_status = false;
+        }
+        return ExecutionControl::Continue;
+    case Names::ContentLength:
+        if (!content_length.has_value()) {
+            content_length = 0;
+        }
+        *content_length = 10 * *content_length + (event.payload - '0');
+        return ExecutionControl::Continue;
+    case Names::Version:
+        switch (event.payload) {
+        case '.':
+            version_major = version_minor;
+            version_minor = 0;
+            return ExecutionControl::Continue;
+        case '0' ... '9':
```

Comment:
I assume you know the case's range syntax (0 ... 9) is a GCC extension.

Reply:
Not really (now I do). Should I replace it? I guess we don't really compile under anything else.

Reply:
Generally we prefer writing standard-compliant code. Probably clang understands this construct as well as QtCreator's code model was not complaining when I was reviewing the code.
In this case, I'd keep your syntax. I just wanted to make sure we agree that this is a potentially non-portable extension.